File: 2015/julia/08.jl
# Advent of Code 2015, day 8
# (c) aichingert

module Day8

function part1(lines)
	ans = 0

	for i in eachindex(lines)
		j = 1
		c = 0

		while j < length(lines[i])
			if lines[i][j] == '\\'
				if '\\' == lines[i][j+1] || '"' == lines[i][j+1]
					j += 1
					c += 1
				elseif 'x' == lines[i][j+1]
					j += 3
					c += 1
				end
			elseif '"' != lines[i][j] c += 1 end

			j += 1
		end

		ans += length(lines[i]) - c
	end

	return ans
end

function part2(lines)
	ans = 0

	for i in eachindex(lines)
		c = 2
		for j in eachindex(lines[i])
			if lines[i][j] == '"' || lines[i][j] == '\\' c += 2
			else c += 1 end
		end

		ans += c - length(lines[i])
	end

	return ans
end

input = readlines("../input/08")

println("Part 1: " * string(part1(input)))
println("Part 2: " * string(part2(input)))

end

File: 2017/rust/08.rs
// Advent of Code 2017, day 8
// (c) aichingert

use std::collections::HashMap;

fn solve() -> (i32,i32) {
    let mut sol: (i32,i32) = (0,0);
    let mut reg: HashMap<&str, i32> = HashMap::new();
    let lines = std::fs::read_to_string("../input/08").unwrap();

    for line in lines.lines() {
        let vals = line.split(' ').collect::<Vec<&str>>();
        
        let cond = match vals[5] {
            ">" => *reg.entry(vals[4]).or_insert(0) > vals[6].parse().unwrap(),
            "<" => *reg.entry(vals[4]).or_insert(0) < vals[6].parse().unwrap(),
            ">="=> *reg.entry(vals[4]).or_insert(0) >= vals[6].parse().unwrap(),
            "<="=> *reg.entry(vals[4]).or_insert(0) <= vals[6].parse().unwrap(),
            "=="=> *reg.entry(vals[4]).or_insert(0) == vals[6].parse().unwrap(),
            "!="=> *reg.entry(vals[4]).or_insert(0) != vals[6].parse().unwrap(),
            _ => panic!("invalid operation")
        };

        if cond {
            update(&mut reg, &vals, &mut sol);
        }
    }

    for key in reg.keys() {
        sol.0 = sol.0.max(reg[key]);
    }

    sol
}

fn update<'a>(reg: &mut HashMap<&'a str, i32>, vals: &Vec<&'a str>, sol: &mut (i32,i32)) {
    let cur = reg.entry(vals[0]).or_insert(0);

    match vals[1] {
        "inc" => *cur += vals[2].parse::<i32>().unwrap(),
        "dec" => *cur -= vals[2].parse::<i32>().unwrap(),
        _ => unreachable!()
    };

    sol.1 = sol.1.max(*cur);
}

fn main() {
    let (part1,part2) = solve();

    println!("Part 1: {}", part1);
    println!("Part 2: {}", part2);
}
